nvbench/cupti_profiler.cpp

```diff
@@ -0,0 +1,790 @@
+    params.pBuffer        = m_data_image_prefix.data();
+
+    nvpw_call(NVPW_CounterDataBuilder_GetCounterDataPrefix(&params));
+  }
+}
+
+[[nodiscard]] std::size_t
+get_counter_data_image_size(CUpti_Profiler_CounterDataImageOptions *options)
+{
+  CUpti_Profiler_CounterDataImage_CalculateSize_Params params = {
+    CUpti_Profiler_CounterDataImage_CalculateSize_Params_STRUCT_SIZE};
+
+  params.pOptions = options;
+  params.sizeofCounterDataImageOptions =
+    CUpti_Profiler_CounterDataImageOptions_STRUCT_SIZE;
+
+  cupti_call(cuptiProfilerCounterDataImageCalculateSize(&params));
+  return params.counterDataImageSize;
+}
+
+
+void cupti_profiler::initialize_counter_data_image()
+{
+  CUpti_Profiler_CounterDataImageOptions counter_data_image_options;
+
+  counter_data_image_options.pCounterDataPrefix = &m_data_image_prefix[0];
+  counter_data_image_options.counterDataPrefixSize = m_data_image_prefix.size();
+  counter_data_image_options.maxNumRanges = m_num_ranges;
+  counter_data_image_options.maxNumRangeTreeNodes = m_num_ranges;
+  counter_data_image_options.maxRangeNameLength = 64;
+
+  m_data_image.resize(
+    get_counter_data_image_size(&counter_data_image_options));
+
+  {
+    CUpti_Profiler_CounterDataImage_Initialize_Params params = {
+      CUpti_Profiler_CounterDataImage_Initialize_Params_STRUCT_SIZE};
+
+    params.sizeofCounterDataImageOptions =
+      CUpti_Profiler_CounterDataImageOptions_STRUCT_SIZE;
+    params.pOptions             = &counter_data_image_options;
+    params.counterDataImageSize = m_data_image.size();
+
+    params.pCounterDataImage = &m_data_image[0];
+    cupti_call(cuptiProfilerCounterDataImageInitialize(&params));
+  }
+
+  {
+    CUpti_Profiler_CounterDataImage_CalculateScratchBufferSize_Params params = {
+      CUpti_Profiler_CounterDataImage_CalculateScratchBufferSize_Params_STRUCT_SIZE};
+
+    params.counterDataImageSize = m_data_image.size();
+    params.pCounterDataImage    = &m_data_image[0];
+
+    cupti_call(
+      cuptiProfilerCounterDataImageCalculateScratchBufferSize(&params));
+
+    m_data_scratch_buffer.resize(params.counterDataScratchBufferSize);
+  }
+
+  {
+    CUpti_Profiler_CounterDataImage_InitializeScratchBuffer_Params params = {
+      CUpti_Profiler_CounterDataImage_InitializeScratchBuffer_Params_STRUCT_SIZE};
+
+    params.counterDataImageSize         = m_data_image.size();
+    params.pCounterDataImage            = &m_data_image[0];
+    params.counterDataScratchBufferSize = m_data_scratch_buffer.size();
+    params.pCounterDataScratchBuffer    = &m_data_scratch_buffer[0];
+
+    cupti_call(cuptiProfilerCounterDataImageInitializeScratchBuffer(&params));
+  }
+}
+
+
+cupti_profiler::~cupti_profiler()
+{
+  if (initialized())
+  {
+    CUpti_Profiler_DeInitialize_Params params = {
+      CUpti_Profiler_DeInitialize_Params_STRUCT_SIZE};
+
+    cuptiProfilerDeInitialize(&params);
+  }
+}
+
+
+[[nodiscard]] bool cupti_profiler::initialized() const
```

Comment:
I'm not sure I understand. Which case are you referring to? https://godbolt.org/z/rd8zfcMKW 
Neither bool nor std::vector has `[[nodiscard]]` in their declarations. Having `[[nodiscard]]` in the `cupti_profiler` wouldn't affect it's methods.

Reply:
The `[[nodiscard]]` attribute is already used inside the definition of `cupti_profiler` on the declaration of the `cupti_profiler::initialize` method:

```
struct cupti_profiler
{
  // ...
  [[nodiscard]] bool initialized() const;
  // ...
};
```

and it's not necessary to repeat it here on the method definition. Ex: https://godbolt.org/z/KMT8hhW6v

I prefer to just have these attributes inside the class definition to avoid having to keep both decl/def sync'd, but it's admittedly an extremely minor nitpick 🙂